app/code/Magento/InventorySales/Model/ReturnProcessor/ProcessRefundItems.php

```diff
@@ -132,6 +141,28 @@ public function execute(
                 $qtyBackToSource -= $backQty;
             }
 
+            foreach ($invoiceItems as $sourceCode => $data) {
```

Comment:
this code fully duplicates logic for products with weight 
Now it looks like 
```php
            foreach ($shippedItems as $sourceCode => $data) {
                if (empty($data[$sku])) {
                    continue;
                }

                $availableQtyToReturn = $data[$sku] + $originalProcessedQty;
                $backQty = min($availableQtyToReturn, $qtyBackToSource);
                $originalProcessedQty += $data[$sku];

                if ($this->isZero((float)$availableQtyToReturn)) {
                    continue;
                }

                $backItemQtyRequest = $this->backItemQtyRequestFactory->create([
                    'sourceCode' => $sourceCode,
                    'sku' => $sku,
                    'qty' => $backQty
                ]);
                $this->processBackItemQtyToSource->execute($backItemQtyRequest, $salesChannel);
                $qtyBackToSource -= $backQty;
            }

            foreach ($invoiceItems as $sourceCode => $data) {
                if (empty($data[$sku])) {
                    continue;
                }

                $availableQtyToReturn = $data[$sku] + $originalProcessedQty;
                $backQty = min($availableQtyToReturn, $qtyBackToSource);
                $originalProcessedQty += $data[$sku];

                if ($this->isZero((float)$availableQtyToReturn)) {
                    continue;
                }

                $backItemQtyRequest = $this->backItemQtyRequestFactory->create([
                    'sourceCode' => $sourceCode,
                    'sku' => $sku,
                    'qty' => $backQty
                ]);
                $this->processBackItemQtyToSource->execute($backItemQtyRequest, $salesChannel);
                $qtyBackToSource -= $backQty;
            }
```
moreover both pieces of logic looks identically, the difference is whether it manipulate with 
`shippedItems` or `invoiceItems`.

Also, currently, both of these processings go one by another, but in real life there is no situation when we need to process both, as 
we need to process `shippedItems` - for products with weight
we need to process `invoiceItems` - for products without weight (virtual & downloadable)

not AND but OR, depending on product type. This means that we need to use 
`\Magento\InventoryConfigurationApi\Model\IsSourceItemManagementAllowedForProductTypeInterface` to determine which one to call.

Also, we need to fix copy-paste as the logic of processing should be impelmented in the dedicated private method which accepts whether `shippedItems` or `invoiceItems` and iteraes over them.

Fix naming:
`$availableQtyToBack`-> `$availableQtyToReturn`

names should be similar, thus if you have `shippedItems`, you also need to call other items as `invoiceDItems`